File: apps/backend/ModuAgent/mcp/discovery.py
"""MCP 工具发现与缓存。

从 MCP Server 的 ``tools/list`` 响应解析工具元信息，
提供缓存、查询能力，供 MCPToolAdapter 使用。
"""

from __future__ import annotations

import logging
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional

logger = logging.getLogger(__name__)


@dataclass
class ToolInfo:
    """MCP 工具元信息（从 tools/list 响应解析）。

    字段对应 MCP 规范的 Tool 定义。

    Attributes:
        server_name: 来源 Server 名
        raw_name: Server 内工具名
        description: 工具描述
        input_schema: JSON Schema 参数定义
    """

    server_name: str
    raw_name: str
    description: str = ""
    input_schema: Dict[str, Any] = field(default_factory=dict)

    @property
    def qualified_name(self) -> str:
        """全限定名：``server_name__raw_name``（避免跨 Server 工具名冲突）。"""
        return f"{self.server_name}__{self.raw_name}"

    @classmethod
    def from_mcp_dict(cls, server_name: str, raw: Dict[str, Any]) -> ToolInfo:
        """从 MCP ``tools/list`` 响应项构建 ToolInfo。

        Args:
            server_name: 来源 Server 名
            raw: MCP 响应中的单个工具字典

        Returns:
            ToolInfo 实例
        """
        return cls(
            server_name=server_name,
            raw_name=raw.get("name", ""),
            description=raw.get("description", ""),
            input_schema=raw.get("inputSchema", {}) or raw.get("input_schema", {}),
        )

    def to_base_tool_schema(self) -> Dict[str, Any]:
        """转换为 ModuAgent ``BaseTool.parameters_schema()`` 格式。

        MCP 的 inputSchema 已是标准 JSON Schema，
        直接返回即可。
        """
        return self.input_schema if self.input_schema else {
            "type": "object",
            "properties": {},
            "additionalProperties": True,
        }


class ToolDiscovery:
    """工具发现服务。

    提供工具发现、缓存、查询能力。
    由 MCPClient 调用，不直接持有 transport。
    """

    def __init__(self) -> None:
        self._cache: Dict[str, List[ToolInfo]] = {}  # server_name → tools

    def update(self, server_name: str, tools: List[ToolInfo]) -> None:
        """更新指定 Server 的工具缓存。

        Args:
            server_name: Server 名
            tools: 工具列表
        """
        self._cache[server_name] = tools
        logger.info("Tool cache updated: server=%s, count=%d", server_name, len(tools))

    def get_all(self) -> List[ToolInfo]:
        """返回所有缓存工具。

        Returns:
            全部 Server 的工具列表
        """
        all_tools: List[ToolInfo] = []
        for tools in self._cache.values():
            all_tools.extend(tools)
        return all_tools

    def get_by_server(self, server_name: str) -> List[ToolInfo]:
        """返回指定 Server 的工具。

        Args:
            server_name: Server 名

        Returns:
            该 Server 的工具列表（可能为空）
        """
        return self._cache.get(server_name, [])

    def find_by_name(self, tool_name: str) -> Optional[ToolInfo]:
        """按全限定名或裸名查找工具。

        优先匹配全限定名，其次裸名（首个命中）。

        Args:
            tool_name: 全限定名或裸名

        Returns:
            ToolInfo 实例，或 None
        """
        # 全限定名匹配
        for tools in self._cache.values():
            for tool in tools:
                if tool.qualified_name == tool_name:
                    return tool
        # 裸名匹配
        for tools in self._cache.values():
            for tool in tools:
                if tool.raw_name == tool_name:
                    return tool
        return None

    def clear(self) -> None:
        """清空缓存。"""
        self._cache.clear()

File: packages/modu-agent/src/mcp/discovery.ts
// 对应 Python: mcp/discovery.py
// MCP 工具发现与缓存。
//
// 从 MCP Server 的 `tools/list` 响应解析工具元信息，
// 提供缓存、查询能力，供 MCPToolAdapter 使用。

const logger = {
  info: (msg: string, ...args: any[]) => console.info(`[mcp] ${msg}`, ...args),
  warning: (msg: string, ...args: any[]) => console.warn(`[mcp] ${msg}`, ...args),
  error: (msg: string, ...args: any[]) => console.error(`[mcp] ${msg}`, ...args),
  debug: (msg: string, ...args: any[]) => console.debug(`[mcp] ${msg}`, ...args),
}

/**
 * MCP 工具元信息（从 tools/list 响应解析）。
 *
 * 字段对应 MCP 规范的 Tool 定义。
 *
 * 对应 Python ToolInfo dataclass。
 */
export class ToolInfo {
  /** 来源 Server 名 */
  serverName: string
  /** Server 内工具名 */
  rawName: string
  /** 工具描述 */
  description: string
  /** JSON Schema 参数定义 */
  inputSchema: Record<string, any>

  constructor(
    serverName: string,
    rawName: string,
    description: string = '',
    inputSchema: Record<string, any> = {},
  ) {
    this.serverName = serverName
    this.rawName = rawName
    this.description = description
    this.inputSchema = inputSchema
  }

  /** 全限定名：`server_name__raw_name`（避免跨 Server 工具名冲突）。 */
  get qualifiedName(): string {
    return `${this.serverName}__${this.rawName}`
  }

  /**
   * 从 MCP `tools/list` 响应项构建 ToolInfo。
   *
   * 对应 Python ToolInfo.from_mcp_dict。
   *
   * @param serverName - 来源 Server 名
   * @param raw - MCP 响应中的单个工具字典
   * @returns ToolInfo 实例
   */
  static fromMcpDict(serverName: string, raw: Record<string, any>): ToolInfo {
    return new ToolInfo(
      serverName,
      raw.name ?? '',
      raw.description ?? '',
      raw.inputSchema ?? raw.input_schema ?? {},
    )
  }

  /**
   * 转换为 ModuAgent `BaseTool.parametersSchema()` 格式。
   *
   * MCP 的 inputSchema 已是标准 JSON Schema，直接返回即可。
   *
   * 对应 Python ToolInfo.to_base_tool_schema。
   */
  toBaseToolSchema(): Record<string, any> {
    if (this.inputSchema && Object.keys(this.inputSchema).length > 0) {
      return this.inputSchema
    }
    return {
      type: 'object',
      properties: {},
      additionalProperties: true,
    }
  }
}

/**
 * 工具发现服务。
 *
 * 提供工具发现、缓存、查询能力。
 * 由 MCPClient 调用，不直接持有 transport。
 *
 * 对应 Python ToolDiscovery。
 */
export class ToolDiscovery {
  /** server_name → tools */
  private _cache: Map<string, ToolInfo[]> = new Map()

  /**
   * 更新指定 Server 的工具缓存。
   *
   * 对应 Python ToolDiscovery.update。
   */
  update(serverName: string, tools: ToolInfo[]): void {
    this._cache.set(serverName, tools)
    logger.info('Tool cache updated: server=%s, count=%d', serverName, tools.length)
  }

  /**
   * 返回所有缓存工具。
   *
   * 对应 Python ToolDiscovery.get_all。
   */
  getAll(): ToolInfo[] {
    const allTools: ToolInfo[] = []
    for (const tools of this._cache.values()) {
      allTools.push(...tools)
    }
    return allTools
  }

  /**
   * 返回指定 Server 的工具。
   *
   * 对应 Python ToolDiscovery.get_by_server。
   */
  getByServer(serverName: string): ToolInfo[] {
    return this._cache.get(serverName) ?? []
  }

  /**
   * 按全限定名或裸名查找工具。
   *
   * 优先匹配全限定名，其次裸名（首个命中）。
   *
   * 对应 Python ToolDiscovery.find_by_name。
   */
  findByName(toolName: string): ToolInfo | null {
    // 全限定名匹配
    for (const tools of this._cache.values()) {
      for (const tool of tools) {
        if (tool.qualifiedName === toolName) {
          return tool
        }
      }
    }
    // 裸名匹配
    for (const tools of this._cache.values()) {
      for (const tool of tools) {
        if (tool.rawName === toolName) {
          return tool
        }
      }
    }
    return null
  }

  /** 清空缓存。 */
  clear(): void {
    this._cache.clear()
  }
}
